// libreact/lib/ExitSensor/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var ExitSensor = (function (_super) {
    tslib_1.__extends(ExitSensor, _super);
    function ExitSensor() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.element = null;
        _this.nextElement = null;
        _this.exitInFlight = false;
        _this.exitFinish = false;
        _this.transition = function () {
            _this.exitFinish = true;
            _this.forceUpdate();
        };
        return _this;
    }
    ExitSensor.prototype.componentWillUnmount = function () {
        clearTimeout(this.timeout);
    };
    ExitSensor.prototype.render = function () {
        if (process.env.NODE_ENV !== 'production') {
            var children = this.props.children;
            if (children && (typeof children !== 'object')) {
                console.error(new TypeError('ExitSensors expects a single ReactElement node as a child.'));
            }
        }
        if (this.exitFinish) {
            this.exitFinish = false;
            this.exitInFlight = false;
            this.element = this.nextElement;
            return this.element || null;
        }
        var element = this.props.children;
        if (this.exitInFlight) {
            this.nextElement = element;
            return React.cloneElement(this.element, {
                exiting: true
            });
        }
        if (!this.element) {
            this.element = element;
            return element || null;
        }
        if (!element || (this.element.key !== element.key)) {
            this.exitInFlight = true;
            this.nextElement = element;
            this.timeout = setTimeout(this.transition, this.props.time);
            return React.cloneElement(this.element, {
                exiting: true
            });
        }
        return element || null;
    };
    ExitSensor.defaultProps = {
        time: 200
    };
    return ExitSensor;
}(React.Component));
exports.ExitSensor = ExitSensor;
