// libreact/lib/Dimmer/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = require("react");
var util_1 = require("../util");
var react_universal_interface_1 = require("react-universal-interface");
var Dimmer = (function (_super) {
    tslib_1.__extends(Dimmer, _super);
    function Dimmer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.el = null;
        _this.ref = function (el) {
            _this.el = el;
        };
        return _this;
    }
    Dimmer.prototype.componentDidMount = function () {
        var parent = this.el.parentElement;
        var position = parent.style.getPropertyValue('position');
        if (process.env.NODE_ENV !== 'production') {
            if (position && (position !== 'relative')) {
                console.warn('The "position" style property of a parent element of <Dimmer> must be ' +
                    ("\"relative\" or not set, but \"" + position + "\" was detected. It will be overwritten ") +
                    'to "relative".');
            }
        }
        if (position !== 'relative') {
            parent.style.setProperty('position', 'relative', 'important');
        }
    };
    Dimmer.prototype.render = function () {
        var hidden = this.props.hidden;
        var style = {
            background: this.props.color,
            position: 'absolute',
            top: 0,
            left: 0,
            width: '100%',
            height: '100%',
            zIndex: 2147483647,
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
            transition: "opacity " + this.props.ms + "ms",
        };
        if (hidden) {
            style.opacity = 0;
            style.pointerEvents = 'none';
        }
        return util_1.h('div', {
            ref: this.ref,
            style: style,
        }, hidden ? null : react_universal_interface_1.render(this.props, this.state));
    };
    Dimmer.defaultProps = {
        color: 'rgba(0,0,0,0.5)',
        ms: 300,
    };
    return Dimmer;
}(react_1.Component));
exports.Dimmer = Dimmer;
