// @react-google-maps/api/lib/components/kml/KmlLayer.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/* global google */
import { PureComponent } from 'react';
import { unregisterEvents, applyUpdatersToPropsAndRegisterEvents } from '../../utils/helper';
import MapContext from '../../map-context';
var eventMap = {
    onClick: 'click',
    onDefaultViewportChanged: 'defaultviewport_changed',
    onStatusChanged: 'status_changed'
};
var updaterMap = {
    options: function (instance, options) {
        instance.setOptions(options);
    },
    url: function (instance, url) {
        instance.setUrl(url);
    },
    zIndex: function (instance, zIndex) {
        instance.setZIndex(zIndex);
    }
};
var KmlLayer = /** @class */ (function (_super) {
    __extends(KmlLayer, _super);
    function KmlLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.registeredEvents = [];
        _this.state = {
            kmlLayer: null
        };
        _this.componentDidMount = function () {
            var kmlLayer = new google.maps.KmlLayer(_this.props.options);
            _this.setState(function () { return ({
                kmlLayer: kmlLayer
            }); }, function () {
                _this.state.kmlLayer.setMap(_this.context);
                _this.registeredEvents = applyUpdatersToPropsAndRegisterEvents({
                    updaterMap: updaterMap,
                    eventMap: eventMap,
                    prevProps: {},
                    nextProps: _this.props,
                    instance: _this.state.kmlLayer
                });
            });
        };
        _this.componentDidUpdate = function (prevProps) {
            unregisterEvents(_this.registeredEvents);
            _this.registeredEvents = applyUpdatersToPropsAndRegisterEvents({
                updaterMap: updaterMap,
                eventMap: eventMap,
                prevProps: prevProps,
                nextProps: _this.props,
                instance: _this.state.kmlLayer
            });
        };
        _this.componentWillUnmount = function () {
            unregisterEvents(_this.registeredEvents);
            _this.state.kmlLayer && _this.state.kmlLayer.setMap(null);
        };
        _this.render = function () { return null; };
        _this.getDefaultViewport = function () { return _this.state.kmlLayer.getDefaultViewport(); };
        _this.getMap = function () { return _this.state.kmlLayer.getMap(); };
        _this.getMetadata = function () { return _this.state.kmlLayer.getMetadata(); };
        _this.getStatus = function () { return _this.state.kmlLayer.getStatus(); };
        _this.getUrl = function () { return _this.state.kmlLayer.getUrl(); };
        _this.getZIndex = function () { return _this.state.kmlLayer.getZIndex(); };
        return _this;
    }
    KmlLayer.contextType = MapContext;
    return KmlLayer;
}(PureComponent));
export { KmlLayer };
export default KmlLayer;
