// libreact/lib/ErrorBoundary/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = require("react");
var util_1 = require("../util");
var ErrorBoundary = (function (_super) {
    tslib_1.__extends(ErrorBoundary, _super);
    function ErrorBoundary() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {};
        return _this;
    }
    ErrorBoundary.prototype.componentDidCatch = function (error, info) {
        this.setState({
            error: error,
            info: info
        });
        (this.props.onError || util_1.noop)(error, info);
    };
    ErrorBoundary.prototype.render = function () {
        var _a = this, props = _a.props, state = _a.state;
        var renderError = props.renderError, children = props.children;
        return state.error ?
            (typeof renderError === 'function' ? renderError(state) : null) :
            children;
    };
    return ErrorBoundary;
}(react_1.Component));
exports.ErrorBoundary = ErrorBoundary;
exports.withErrorBoundary = function (Comp, boundaryProps) { return function (props) {
    return util_1.h(ErrorBoundary, boundaryProps, util_1.h(Comp, props));
}; };
