// libreact/lib/Flipflop/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("../util");
var State_1 = require("../State");
var renderProp_1 = tslib_1.__importDefault(require("../util/renderProp"));
var faccToHoc_1 = tslib_1.__importDefault(require("../util/faccToHoc"));
exports.Flipflop = function (props) {
    var initialState = props.init || false;
    return util_1.h(State_1.State, {
        init: { on: initialState },
        render: function (_a, set) {
            var on = _a.on;
            return renderProp_1.default(props, {
                on: on,
                flip: function () {
                    if (on === initialState) {
                        set({ on: !initialState });
                    }
                },
                flop: function () {
                    if (on !== initialState) {
                        set({ on: initialState });
                    }
                }
            });
        }
    });
};
exports.withFlipflop = faccToHoc_1.default(exports.Flipflop, 'flipflop');
