// libreact/modules/LocalStorage/index.js
import { Component } from 'react';
import { get, set, del } from './local-storage';
import { debounce } from 'throttle-debounce';
export class LocalStorage extends Component {
    constructor() {
        super(...arguments);
        this.put = debounce(this.props.debounce, (rawData) => {
            const { name, data } = this.props;
            try {
                if (!rawData) {
                    rawData = JSON.stringify(data);
                }
                set(name, rawData);
            }
            catch { }
        });
    }
    componentDidMount() {
        const { name, onMount } = this.props;
        if (onMount) {
            let json = '';
            try {
                json = get(name);
            }
            catch {
                json = 'null';
            }
            if (typeof json === 'string') {
                try {
                    const data = JSON.parse(json);
                    onMount(data);
                }
                catch { }
            }
        }
        else {
            this.put();
        }
    }
    componentDidUpdate(props) {
        if (props.name !== this.props.name) {
            this.remove();
            this.put();
        }
        else {
            try {
                const newJson = JSON.stringify(this.props.data);
                if (JSON.stringify(props.data) !== newJson) {
                    this.put(newJson);
                }
            }
            catch { }
        }
    }
    componentWillUnmount() {
        this.remove();
    }
    remove(name = this.props.name) {
        if (!this.props.persist) {
            try {
                del(name);
            }
            catch { }
        }
    }
    render() {
        return null;
    }
}
LocalStorage.defaultProps = {
    debounce: 200
};
