// libreact/lib/AfterTimeout/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = require("react");
var AfterTimeout = (function (_super) {
    tslib_1.__extends(AfterTimeout, _super);
    function AfterTimeout() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            ready: false
        };
        return _this;
    }
    AfterTimeout.prototype.componentDidMount = function () {
        var _this = this;
        this.timer = setTimeout(function () {
            _this.setState({
                ready: true
            });
        }, this.props.ms);
    };
    AfterTimeout.prototype.componentWillUnmount = function () {
        clearTimeout(this.timer);
    };
    AfterTimeout.prototype.render = function () {
        return this.state.ready ? this.props.children : null;
    };
    AfterTimeout.defaultProps = {
        ms: 200
    };
    return AfterTimeout;
}(react_1.Component));
exports.AfterTimeout = AfterTimeout;
