// libreact/lib/ActiveSensor/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = require("react");
var util_1 = require("../util");
var Value_1 = require("../Value");
var renderProp_1 = tslib_1.__importDefault(require("../util/renderProp"));
var faccToHoc_1 = tslib_1.__importDefault(require("../util/faccToHoc"));
exports.ActiveSensor = function (props) {
    var bond = props.bond;
    return Value_1.Value({
        render: function (_a) {
            var value = _a.value, set = _a.set;
            var _b;
            if (bond) {
                if (typeof bond === 'boolean') {
                    bond = 'bond';
                }
                return renderProp_1.default(props, (_b = {
                        isActive: value
                    },
                    _b[bond] = {
                        onMouseDown: function () { return set(true); },
                        onMouseUp: function () { return set(false); },
                        onMouseLeave: function () { return set(false); },
                    },
                    _b));
            }
            else {
                var element = renderProp_1.default(props, {
                    isActive: value
                });
                return react_1.cloneElement(element, {
                    onMouseDown: function () { return set(true); },
                    onMouseUp: function () { return set(false); },
                    onMouseLeave: function () { return set(false); },
                });
            }
        }
    });
};
var ActiveSensorWithBond = function (props) { return util_1.h(exports.ActiveSensor, tslib_1.__assign({ bond: true }, props)); };
exports.withActive = faccToHoc_1.default(ActiveSensorWithBond, 'active');
