// libreact/lib/WhenIdle/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = require("react");
var RIC = window.requestIdleCallback || (function (fn, _a) {
    var _b = (_a === void 0 ? {} : _a).timeout, timeout = _b === void 0 ? 35 : _b;
    return setTimeout(fn, timeout);
});
var WhenIdle = (function (_super) {
    tslib_1.__extends(WhenIdle, _super);
    function WhenIdle() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.mounted = false;
        _this.state = {
            ready: false
        };
        return _this;
    }
    WhenIdle.prototype.componentDidMount = function () {
        var _this = this;
        this.mounted = true;
        var timeout = this.props.timeout;
        RIC(function () {
            if (_this.mounted) {
                _this.setState({ ready: true });
            }
        }, { timeout: timeout });
    };
    WhenIdle.prototype.componentWillUnmount = function () {
        this.mounted = false;
    };
    WhenIdle.prototype.render = function () {
        return this.state.ready ? this.props.children : null;
    };
    return WhenIdle;
}(react_1.Component));
exports.WhenIdle = WhenIdle;
