// libreact/lib/LightSensor/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = require("react");
var faccToHoc_1 = tslib_1.__importDefault(require("../util/faccToHoc"));
var renderProp_1 = tslib_1.__importDefault(require("../util/renderProp"));
var LightSensor = (function (_super) {
    tslib_1.__extends(LightSensor, _super);
    function LightSensor() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            illuminance: NaN
        };
        _this.onChange = function () {
            var illuminance = _this.sensor.illuminance;
            _this.setState({
                illuminance: illuminance
            });
        };
        _this.onError = function (event) {
            var name = event.name, message = event.message;
            _this.setState({
                error: {
                    name: name,
                    message: message
                }
            });
        };
        return _this;
    }
    LightSensor.prototype.componentDidMount = function () {
        if (window.AmbientLightSensor) {
            this.sensor = new AmbientLightSensor();
            this.sensor.onreading = this.onChange;
            this.sensor.onerror = this.onError;
            this.sensor.start();
        }
    };
    LightSensor.prototype.render = function () {
        return renderProp_1.default(this.props, this.state);
    };
    return LightSensor;
}(react_1.Component));
exports.LightSensor = LightSensor;
exports.withLight = faccToHoc_1.default(LightSensor);
