// @react-google-maps/api/lib/components/places/StandaloneSearchBox.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/* global google */
import * as React from 'react';
import { createRef, Children, PureComponent } from 'react';
//@ts-ignore
import invariant from 'invariant'; // Do we need this dependency?
import { unregisterEvents, applyUpdatersToPropsAndRegisterEvents } from '../../utils/helper';
import MapContext from '../../map-context';
var eventMap = {
    onPlacesChanged: 'places_changed'
};
var updaterMap = {
    bounds: function (instance, bounds) {
        instance.setBounds(bounds);
    }
};
var StandaloneSearchBox = /** @class */ (function (_super) {
    __extends(StandaloneSearchBox, _super);
    function StandaloneSearchBox(props, context) {
        var _this = _super.call(this, props, context) || this;
        _this.registeredEvents = [];
        _this.state = {
            searchBox: null
        };
        _this.componentDidMount = function () {
            var searchBox = new google.maps.places.SearchBox(_this.containerElement.current.querySelector('input'), _this.props.options);
            _this.setState(function () { return ({
                searchBox: searchBox
            }); }, function () {
                _this.registeredEvents = applyUpdatersToPropsAndRegisterEvents({
                    updaterMap: updaterMap,
                    eventMap: eventMap,
                    prevProps: {},
                    nextProps: _this.props,
                    instance: _this.state.searchBox
                });
            });
        };
        _this.componentDidUpdate = function (prevProps) {
            unregisterEvents(_this.registeredEvents);
            _this.registeredEvents = applyUpdatersToPropsAndRegisterEvents({
                updaterMap: updaterMap,
                eventMap: eventMap,
                prevProps: prevProps,
                nextProps: _this.props,
                instance: _this.state.searchBox
            });
        };
        _this.componentWillUnmount = function () {
            unregisterEvents(_this.registeredEvents);
        };
        _this.render = function () { return React.createElement("div", { ref: _this.containerElement }, Children.only(_this.props.children)); };
        _this.getBounds = function () { return _this.state.searchBox.getBounds(); };
        _this.getPlaces = function () { return _this.state.searchBox.getPlaces(); };
        invariant(google.maps.places, 'Did you include "libraries=places" in the URL?');
        _this.containerElement = createRef();
        return _this;
    }
    StandaloneSearchBox.contextType = MapContext;
    return StandaloneSearchBox;
}(PureComponent));
export default StandaloneSearchBox;
