// libreact/modules/RenderInterval/index.js
import { Component } from 'react';
import { render, createEnhancer } from 'react-universal-interface';
export class RenderInterval extends Component {
    constructor() {
        super(...arguments);
        this.interval = 50;
        this.timeout = null;
        this.timeoutDelay = null;
        this.state = {
            start: 0,
            now: 0,
            value: 0
        };
        this.start = () => {
            const now = Date.now();
            this.interval = 1000 / this.props.fps;
            this.setState({
                start: now,
                now,
                value: 0
            }, () => {
                this.timeout = setTimeout(this.onFrame, this.interval);
            });
        };
        this.onFrame = () => {
            const now = Date.now();
            const value = Math.min((now - this.state.start) / this.props.ms, 1);
            let onState;
            if (value < 1) {
                onState = () => {
                    this.timeout = setTimeout(this.onFrame, this.interval);
                };
            }
            this.setState({
                now,
                value
            }, onState);
        };
    }
    componentDidMount() {
        const { delay } = this.props;
        if (delay) {
            this.timeoutDelay = setTimeout(this.start, delay);
        }
        else {
            this.start();
        }
    }
    componentDidUpdate(props) {
        if (props.fps !== this.props.fps) {
            this.interval = 1000 / this.props.fps;
        }
    }
    componentWillUnmount() {
        clearTimeout(this.timeout);
        clearTimeout(this.timeoutDelay);
    }
    render() {
        return render(this.props, this.state);
    }
}
RenderInterval.defaultProps = {
    fps: 30,
    ms: 300
};
export const withRenderInterval = createEnhancer(RenderInterval, 'renderInterval');
