// libreact/lib/Portal/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var react_dom_1 = require("react-dom");
var util_1 = require("../util");
var Portal = (function (_super) {
    tslib_1.__extends(Portal, _super);
    function Portal(props, context) {
        var _this = _super.call(this, props, context) || this;
        _this.el = null;
        if (util_1.isClient && !props.el) {
            _this.createEl();
            (props.onElement || util_1.noop)(_this.el);
        }
        return _this;
    }
    Portal.prototype.componentWillUpdate = function (props) {
        if (!props.el && !this.el) {
            this.createEl();
        }
    };
    Portal.prototype.componentWillUnmount = function () {
        if (this.el) {
            document.body.removeChild(this.el);
        }
        this.el = null;
    };
    Portal.prototype.createEl = function () {
        this.el = document.createElement('div');
        document.body.appendChild(this.el);
    };
    Portal.prototype.render = function () {
        if (!util_1.isClient) {
            return null;
        }
        var _a = this.props, el = _a.el, children = _a.children;
        return react_dom_1.createPortal(children, el || this.el);
    };
    return Portal;
}(React.Component));
exports.Portal = Portal;
