// libreact/lib/AfterDraf/createSingleRunDraf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var util_1 = require("../util");
var RAF_1 = tslib_1.__importDefault(require("./RAF"));
exports.createSingleRunDraf = util_1.isClient
    ? function () {
        var signelDrafFinished = false;
        return (function (_super) {
            tslib_1.__extends(class_1, _super);
            function class_1(props, context) {
                var _this = _super.call(this, props, context) || this;
                if (util_1.isClient && !signelDrafFinished) {
                    _this.state = {
                        ready: false
                    };
                }
                return _this;
            }
            class_1.prototype.componentDidMount = function () {
                var _this = this;
                if (!signelDrafFinished) {
                    this.frame = RAF_1.default(function () {
                        _this.frame = RAF_1.default(function () {
                            signelDrafFinished = true;
                            _this.setState({ ready: true });
                        });
                    });
                }
            };
            class_1.prototype.componentWillUnmount = function () {
                if (this.frame) {
                    cancelAnimationFrame(this.frame);
                }
            };
            class_1.prototype.render = function () {
                if (!util_1.isClient || signelDrafFinished) {
                    return this.props.children;
                }
                return this.state.ready ? this.props.children : null;
            };
            return class_1;
        }(React.Component));
    }
    : function () { return function (props) { return props.children; }; };
exports.default = exports.createSingleRunDraf;
