// libreact/lib/AfterDraf/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var util_1 = require("../util");
exports.AfterDraf = util_1.isClient
    ? (function (_super) {
        tslib_1.__extends(AfterDraf, _super);
        function AfterDraf(props, context) {
            var _this = _super.call(this, props, context) || this;
            if (util_1.isClient) {
                _this.state = {
                    ready: false
                };
            }
            return _this;
        }
        AfterDraf.prototype.componentDidMount = function () {
            var _this = this;
            var RAF = requestAnimationFrame;
            this.frame = RAF(function () {
                _this.frame = RAF(function () {
                    _this.setState({ ready: true });
                });
            });
        };
        AfterDraf.prototype.componentWillUnmount = function () {
            cancelAnimationFrame(this.frame);
        };
        AfterDraf.prototype.render = function () {
            return this.state.ready ? this.props.children : null;
        };
        ;
        return AfterDraf;
    }(React.Component)) : function (props) { return props.children; };
