// libreact/modules/Media/index.js
import * as React from 'react';
import { h } from '../util';
import parseTimeRanges from './parseTimeRanges';
const defaultState = {
    buffered: [],
    time: 0,
    duration: 0,
    isPlaying: false,
    muted: false,
    volume: 1,
    canPlay: false,
};
export class Media extends React.Component {
    constructor() {
        super(...arguments);
        this.tag = 'video';
        this.el = null;
        this.state = defaultState;
        this.ref = (el) => {
            this.el = el;
        };
        this.lockPlay = false;
        this.play = () => {
            if (this.el && !this.lockPlay) {
                const promise = this.el.play();
                const isPromise = typeof promise === 'object';
                if (isPromise) {
                    this.lockPlay = true;
                    const resetLock = () => {
                        this.lockPlay = false;
                    };
                    promise.then(resetLock, resetLock);
                }
            }
        };
        this.pause = () => {
            if (this.el && !this.lockPlay) {
                this.el.pause();
            }
        };
        this.seek = (time) => {
            if (this.el) {
                time = Math.min(this.state.duration, Math.max(0, time));
                this.el.currentTime = time;
            }
        };
        this.volume = (volume) => {
            if (this.el) {
                volume = Math.min(1, Math.max(0, volume));
                this.el.volume = volume;
                this.setState({
                    volume
                });
            }
        };
        this.mute = () => {
            if (this.el) {
                this.el.muted = true;
            }
        };
        this.unmute = () => {
            if (this.el) {
                this.el.muted = false;
            }
        };
        this.event = (name) => (event) => {
            const handler = this.props[name];
            if (handler) {
                handler(event, this, this.state);
            }
        };
        this.onPlay = (event) => {
            this.change({
                isPlaying: true
            });
            this.event('onPlay')(event);
        };
        this.onPause = (event) => {
            this.change({
                isPlaying: false
            });
            this.event('onPause')(event);
        };
        this.onVolumeChange = (event) => {
            const { muted, volume } = this.el;
            this.change({
                muted,
                volume
            });
            this.event('onVolumeChange')(event);
        };
        this.onDurationChange = (event) => {
            const { duration, buffered } = this.el;
            this.change({
                duration,
                buffered: parseTimeRanges(buffered)
            });
            this.event('onDurationChange')(event);
        };
        this.onTimeUpdate = (event) => {
            this.change({
                time: this.el.currentTime
            });
            this.event('onTimeUpdate')(event);
        };
        this.onProgress = (event) => {
            this.change({
                buffered: parseTimeRanges(this.el.buffered)
            });
            this.event('onProgress')(event);
        };
        this.onCanPlay = (event) => {
            this.change({
                canPlay: true,
            });
            this.event('onCanPlay')(event);
        };
    }
    componentDidMount() {
        this.setState({
            volume: this.el.volume,
            muted: this.el.muted,
        });
        this.event('onMount')(this);
        if (this.props.autoPlay && this.el.paused) {
            this.play();
        }
    }
    componentWillUnmount() {
        this.el = null;
        this.event('onUnmount')(this);
    }
    componentDidUpdate(prevProps) {
        if (prevProps.src !== this.props.src) {
            this.setState(defaultState);
        }
    }
    change(nextState) {
        this.setState(nextState, () => {
            this.event('onChange')(null);
        });
    }
    render() {
        const { props, event } = this;
        const { tag = this.tag, children, render, noJs, onMount, onUnmount, ...rest } = props;
        return h(tag, {
            controls: false,
            ...rest,
            ref: this.ref,
            onAbort: event('onAbort'),
            onCanPlay: this.onCanPlay,
            onCanPlayThrough: event('onCanPlayThrough'),
            onDurationChange: this.onDurationChange,
            onEmptied: event('onEmptied'),
            onEncrypted: event('onEncrypted'),
            onEnded: event('onEnded'),
            onError: event('onError'),
            onLoadedData: event('onLoadedData'),
            onLoadedMetadata: event('onLoadedMetadata'),
            onLoadStart: event('onLoadStart'),
            onPause: this.onPause,
            onPlay: this.onPlay,
            onPlaying: event('onPlaying'),
            onProgress: this.onProgress,
            onRateChange: event('onRateChange'),
            onSeeked: event('onSeeked'),
            onSeeking: event('onSeeking'),
            onStalled: event('onStalled'),
            onSuspend: event('onSuspend'),
            onTimeUpdate: this.onTimeUpdate,
            onVolumeChange: this.onVolumeChange,
            onWaiting: event('onWaiting')
        }, noJs);
    }
}
