// libreact/lib/Resolve/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var Resolve = (function (_super) {
    tslib_1.__extends(Resolve, _super);
    function Resolve(props, context) {
        var _this = _super.call(this, props, context) || this;
        _this.mounted = false;
        _this.state = {
            pending: true
        };
        _this.resolve();
        return _this;
    }
    Resolve.prototype.componentDidMount = function () {
        this.mounted = true;
    };
    Resolve.prototype.componentDidUpdate = function (_a) {
        var promise = _a.promise;
        if (promise !== this.props.promise) {
            this.resolve();
        }
    };
    Resolve.prototype.componentWillUnmount = function () {
        this.mounted = false;
    };
    Resolve.prototype.resolve = function () {
        var _this = this;
        var promise = this.props.promise;
        promise
            .then(function (value) {
            if (_this.mounted && (promise === _this.props.promise)) {
                _this.setState({
                    pending: false,
                    value: value
                });
            }
        })
            .catch(function (error) {
            if (_this.mounted && (promise === _this.props.promise)) {
                _this.setState({
                    pending: false,
                    error: error
                });
            }
        });
    };
    Resolve.prototype.render = function () {
        return this.props.children(this.state);
    };
    return Resolve;
}(React.Component));
exports.Resolve = Resolve;
