// @react-google-maps/api/lib/components/directions/DirectionsRenderer.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/* global google */
import { PureComponent } from 'react';
import { unregisterEvents, applyUpdatersToPropsAndRegisterEvents } from '../../utils/helper';
import MapContext from '../../map-context';
var eventMap = {
    onDirectionsChanged: 'directions_changed'
};
var updaterMap = {
    directions: function (instance, directions) {
        instance.setDirections(directions);
    },
    map: function (instance, map) {
        instance.setMap(map);
    },
    options: function (instance, options) {
        instance.setOptions(options);
    },
    panel: function (instance, panel) {
        instance.setPanel(panel);
    },
    routeIndex: function (instance, routeIndex) {
        instance.setRouteIndex(routeIndex);
    }
};
var DirectionsRenderer = /** @class */ (function (_super) {
    __extends(DirectionsRenderer, _super);
    function DirectionsRenderer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.registeredEvents = [];
        _this.state = {
            directionsRenderer: null
        };
        _this.componentDidMount = function () {
            var directionsRenderer = new google.maps.DirectionsRenderer(_this.props.options);
            _this.setState(function () { return ({
                directionsRenderer: directionsRenderer
            }); }, function () {
                _this.state.directionsRenderer.setMap(_this.context);
                _this.registeredEvents = applyUpdatersToPropsAndRegisterEvents({
                    updaterMap: updaterMap,
                    eventMap: eventMap,
                    prevProps: {},
                    nextProps: _this.props,
                    instance: _this.state.directionsRenderer
                });
            });
        };
        _this.componentDidUpdate = function (prevProps) {
            unregisterEvents(_this.registeredEvents);
            _this.registeredEvents = applyUpdatersToPropsAndRegisterEvents({
                updaterMap: updaterMap,
                eventMap: eventMap,
                prevProps: prevProps,
                nextProps: _this.props,
                instance: _this.state.directionsRenderer
            });
        };
        _this.componentWillUnmount = function () {
            unregisterEvents(_this.registeredEvents);
            if (_this.state.directionsRenderer) {
                _this.state.directionsRenderer.setMap(null);
            }
        };
        _this.render = function () { return null; };
        _this.getDirections = function () { return _this.state.directionsRenderer.getDirections(); };
        _this.getMap = function () { return _this.state.directionsRenderer.getMap(); };
        _this.getPanel = function () { return _this.state.directionsRenderer.getPanel(); };
        _this.getRouteIndex = function () { return _this.state.directionsRenderer.getRouteIndex(); };
        return _this;
    }
    DirectionsRenderer.contextType = MapContext;
    return DirectionsRenderer;
}(PureComponent));
export { DirectionsRenderer };
export default DirectionsRenderer;
