// @react-google-maps/api/lib/components/heatmap/HeatmapLayer.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
/* global google */
import { PureComponent } from 'react';
//@ts-ignore
import invariant from 'invariant';
import { unregisterEvents, applyUpdatersToPropsAndRegisterEvents } from '../../utils/helper';
import MapContext from '../../map-context';
var eventMap = {};
var updaterMap = {
    data: function (instance, data) {
        instance.setData(data);
    },
    map: function (instance, map) {
        instance.setMap(map);
    },
    options: function (instance, options) {
        // TODO: add to official typings
        //@ts-ignore
        instance.setOptions(options);
    }
};
var HeatmapLayer = /** @class */ (function (_super) {
    __extends(HeatmapLayer, _super);
    function HeatmapLayer(props) {
        var _this = _super.call(this, props) || this;
        _this.registeredEvents = [];
        _this.state = {
            heatmapLayer: null
        };
        _this.componentDidMount = function () {
            var heatmapLayer = new google.maps.visualization.HeatmapLayer(__assign({}, _this.props.options, { map: _this.context }));
            _this.setState(function () { return ({
                heatmapLayer: heatmapLayer
            }); }, function () {
                _this.registeredEvents = applyUpdatersToPropsAndRegisterEvents({
                    updaterMap: updaterMap,
                    eventMap: eventMap,
                    prevProps: {},
                    nextProps: _this.props,
                    instance: _this.state.heatmapLayer
                });
            });
        };
        _this.componentDidUpdate = function (prevProps) {
            unregisterEvents(_this.registeredEvents);
            _this.registeredEvents = applyUpdatersToPropsAndRegisterEvents({
                updaterMap: updaterMap,
                eventMap: eventMap,
                prevProps: prevProps,
                nextProps: _this.props,
                instance: _this.state.heatmapLayer
            });
        };
        _this.componentWillUnmount = function () {
            unregisterEvents(_this.registeredEvents);
            if (_this.state.heatmapLayer) {
                _this.state.heatmapLayer.setMap(null);
            }
        };
        _this.render = function () { return null; };
        _this.getData = function () { return _this.state.heatmapLayer.getData(); };
        _this.getMap = function () { return _this.state.heatmapLayer.getMap(); };
        invariant(google.maps.visualization, 'Did you include "visualization" in the libraries array prop in <LoadScript />?');
        return _this;
    }
    HeatmapLayer.contextType = MapContext;
    return HeatmapLayer;
}(PureComponent));
export { HeatmapLayer };
export default HeatmapLayer;
