// libreact/lib/Render/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = require("react");
var react_universal_interface_1 = require("react-universal-interface");
var Render = (function (_super) {
    tslib_1.__extends(Render, _super);
    function Render() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.timeout = null;
        _this.frame = null;
        _this.state = {
            start: 0,
            now: 0,
            value: 0
        };
        _this.start = function () {
            var now = Date.now();
            _this.setState({
                start: now,
                now: now,
                value: 0
            }, function () {
                _this.frame = requestAnimationFrame(_this.onFrame);
            });
        };
        _this.onFrame = function () {
            var now = Date.now();
            var value = Math.min((now - _this.state.start) / _this.props.ms, 1);
            var onState;
            if (value < 1) {
                onState = function () {
                    _this.frame = requestAnimationFrame(_this.onFrame);
                };
            }
            _this.setState({
                now: now,
                value: value
            }, onState);
        };
        return _this;
    }
    Render.prototype.componentDidMount = function () {
        var delay = this.props.delay;
        if (delay) {
            this.timeout = setTimeout(this.start, delay);
        }
        else {
            this.start();
        }
    };
    Render.prototype.componentWillUnmount = function () {
        cancelAnimationFrame(this.frame);
        clearTimeout(this.timeout);
    };
    Render.prototype.render = function () {
        return react_universal_interface_1.render(this.props, this.state);
    };
    Render.defaultProps = {
        ms: 300
    };
    return Render;
}(react_1.Component));
exports.Render = Render;
exports.withRender = react_universal_interface_1.createEnhancer(Render, 'render');
