// libreact/lib/ScrollSensor/story.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = require("react");
var react_2 = require("@storybook/react");
var _1 = require(".");
var ShowDocs_1 = tslib_1.__importDefault(require("../ShowDocs"));
var StoryScrollSensorBasic = (function (_super) {
    tslib_1.__extends(StoryScrollSensorBasic, _super);
    function StoryScrollSensorBasic() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            el: null
        };
        _this.ref = function (el) { return _this.setState({ el: el }); };
        return _this;
    }
    StoryScrollSensorBasic.prototype.render = function () {
        return react_1.createElement('div', {
            ref: this.ref,
            style: {
                border: '1px solid red',
                width: 300,
                height: 500,
                overflow: 'scroll'
            }
        }, react_1.createElement(_1.ScrollSensor, { el: this.state.el }, function (_a) {
            var x = _a.x, y = _a.y;
            return react_1.createElement('div', {
                style: {
                    position: 'absolute',
                    top: 20,
                    left: 20
                }
            }, "x: " + x + ", y: " + y);
        }), react_1.createElement('div', { style: {
                background: 'linear-gradient(to bottom, #fff 0%, #000 100%)',
                width: 2000,
                height: 2000
            } }));
    };
    return StoryScrollSensorBasic;
}(react_1.Component));
react_2.storiesOf('Sensors/ScrollSensor', module)
    .add('Documentation', function () { return react_1.createElement(ShowDocs_1.default, { md: require('../../docs/en/ScrollSensor.md') }); })
    .add('Example', function () { return react_1.createElement(StoryScrollSensorBasic); });
