// libreact/lib/translate/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = require("react");
var util_1 = require("../util");
var context_1 = require("../context");
var faccToHoc_1 = tslib_1.__importDefault(require("../util/faccToHoc"));
var renderProp_1 = tslib_1.__importDefault(require("../util/renderProp"));
var Translations = (function (_super) {
    tslib_1.__extends(Translations, _super);
    function Translations() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Translations.prototype.render = function () {
        var _a = this.props, children = _a.children, value = _a.map;
        return util_1.h(context_1.Provider, {
            name: util_1.ns("T/" + this.props.ns),
            value: value
        }, children);
    };
    return Translations;
}(react_1.Component));
exports.Translations = Translations;
var Translate = (function (_super) {
    tslib_1.__extends(Translate, _super);
    function Translate() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Translate.prototype.render = function () {
        var _this = this;
        return util_1.h(context_1.Consumer, { name: util_1.ns("T/" + this.props.ns) }, function (map) {
            var T = function (key) {
                var args = [];
                for (var _i = 1; _i < arguments.length; _i++) {
                    args[_i - 1] = arguments[_i];
                }
                var value = map[key];
                if (typeof value === 'function') {
                    return value.apply(void 0, [T].concat(args));
                }
                return value || key;
            };
            return renderProp_1.default(_this.props, T);
        });
    };
    return Translate;
}(react_1.Component));
exports.Translate = Translate;
exports.T = Translate;
exports.withT = faccToHoc_1.default(Translate, 'T');
