// libreact/lib/route/createRouter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("../util");
var Router_1 = tslib_1.__importDefault(require("./Router"));
var Match_1 = tslib_1.__importDefault(require("./Match"));
var Route_1 = tslib_1.__importDefault(require("./Route"));
var Switch_1 = tslib_1.__importDefault(require("./Switch"));
var go_1 = tslib_1.__importDefault(require("./go"));
var Link_1 = tslib_1.__importDefault(require("./Link"));
var createRouter = function (ns, go) {
    if (go === void 0) { go = go_1.default; }
    var proxy = function (Comp) { return function (props) { return util_1.h(Comp, tslib_1.__assign({}, props, { ns: ns })); }; };
    return {
        Router: proxy(Router_1.default),
        Match: proxy(Match_1.default),
        Route: proxy(Route_1.default),
        Switch: proxy(Switch_1.default),
        go: go,
        Link: function (props) { return util_1.h(Link_1.default, tslib_1.__assign({}, props, { go: go })); },
    };
};
exports.default = createRouter;
