// libreact/lib/route/Match.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var renderProp_1 = tslib_1.__importDefault(require("../util/renderProp"));
var Router_1 = tslib_1.__importDefault(require("./Router"));
var context_1 = require("../context");
var util_1 = require("../util");
var createMatcher_1 = tslib_1.__importDefault(require("./createMatcher"));
var Match = function (props) {
    if (process.env.NODE_ENV !== 'production') {
        if (typeof props.match !== 'string') {
            if (props.exact) {
                console.warn('You are using <Match exact /> with non-string match prop, ' +
                    'exact prop works only with string match prop.');
            }
        }
    }
    return util_1.h(context_1.Consumer, { name: util_1.ns("route/" + props.ns) }, function (context) {
        var exact = props.exact, match = props.match, truncate = props.truncate;
        var fullRoute = context.fullRoute, route = context.route, parent = context.parent;
        var matches = createMatcher_1.default(match, exact)(route);
        var data = {
            fullRoute: fullRoute,
            route: route,
            parent: parent,
            matches: matches,
        };
        var element = renderProp_1.default(props, data);
        if (matches && truncate) {
            var routerProps = {
                fullRoute: route,
                route: route.substr(matches[0].length),
                parent: data,
            };
            element = util_1.h(Router_1.default, routerProps, element);
        }
        return element;
    });
};
Match.defaultProps = {
    match: '',
    exact: false,
};
exports.default = Match;
