// @react-google-maps/api/lib/components/streetview/StreetViewCoverageLayer.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/* global google */
import { PureComponent } from 'react';
import { unregisterEvents, applyUpdatersToPropsAndRegisterEvents } from '../../utils/helper';
import MapContext from '../../map-context';
var eventMap = {};
var updaterMap = {
    map: function (instance, map) {
        instance.setMap(map);
    },
};
var StreetViewCoverageLayer = /** @class */ (function (_super) {
    __extends(StreetViewCoverageLayer, _super);
    function StreetViewCoverageLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.registeredEvents = [];
        _this.state = {
            streetViewCoverageLayer: null
        };
        _this.componentDidMount = function () {
            var streetViewCoverageLayer = new google.maps.StreetViewCoverageLayer();
            _this.setState(function () { return ({
                streetViewCoverageLayer: streetViewCoverageLayer
            }); }, function () {
                _this.state.streetViewCoverageLayer.setMap(_this.context);
                _this.registeredEvents = applyUpdatersToPropsAndRegisterEvents({
                    updaterMap: updaterMap,
                    eventMap: eventMap,
                    prevProps: {},
                    nextProps: _this.props,
                    instance: _this.state.streetViewCoverageLayer
                });
            });
        };
        _this.componentDidUpdate = function (prevProps) {
            unregisterEvents(_this.registeredEvents);
            _this.registeredEvents = applyUpdatersToPropsAndRegisterEvents({
                updaterMap: updaterMap,
                eventMap: eventMap,
                prevProps: prevProps,
                nextProps: _this.props,
                instance: _this.state.streetViewCoverageLayer
            });
        };
        _this.componentWillUnmount = function () {
            unregisterEvents(_this.registeredEvents);
            _this.state.streetViewCoverageLayer &&
                _this.state.streetViewCoverageLayer.setMap(null);
        };
        _this.render = function () { return null; };
        _this.getMap = function () {
            return _this.state.streetViewCoverageLayer.getMap();
        };
        return _this;
    }
    StreetViewCoverageLayer.contextType = MapContext;
    return StreetViewCoverageLayer;
}(PureComponent));
export { StreetViewCoverageLayer };
export default StreetViewCoverageLayer;
